core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -768,13 +692,151 @@ private void sendFailShard(ShardRouting shardRouting, String message, @Nullable
         @Override
         public void handle(final IndexShard.ShardFailure shardFailure) {
-            final IndexService indexService = indicesService.indexService(shardFailure.routing.shardId().getIndex());
             final ShardRouting shardRouting = shardFailure.routing;
             threadPool.generic().execute(() -> {
                 synchronized (mutex) {
-                    failAndRemoveShard(shardRouting, indexService, true, "shard failure, reason [" + shardFailure.reason + "]", shardFailure.cause);
+                    failAndRemoveShard(shardRouting, true, "shard failure, reason [" + shardFailure.reason + "]", shardFailure.cause);
                 }
             });
         }
     }
+
+    public interface Shard {
+
+        /**
+         * Returns the shard id of this shard.
+         */
+        ShardId shardId();
+
+        /**
+         * Returns the latest cluster routing entry received with this shard.
+         */
+        ShardRouting routingEntry();
+
+        /**
+         * Returns the latest internal shard state.
+         */
+        IndexShardState state();
+
+        /**
+         * Returns the recovery state associated with this shard.
+         */
+        RecoveryState recoveryState();
+
+        /**
+         * Updates the shards routing entry. This mutate the shards internal state depending
+         * on the changes that get introduced by the new routing value. This method will persist shard level metadata
+         * unless explicitly disabled.
+         *
+         * @throws IndexShardRelocatedException if shard is marked as relocated and relocation aborted
+         * @throws IOException                  if shard state could not be persisted
+         */
+        void updateRoutingEntry(ShardRouting shardRouting, boolean persistState) throws IOException;
+    }
+
+    public interface AllocatedIndex<T extends Shard> extends Iterable<T>, IndexComponent {
+
+        /**
+         * Returns the index settings of this index.
+         */
+        IndexSettings getIndexSettings();
+
+        /**
+         * Updates the meta data of this index. Changes become visible through {@link #getIndexSettings()}
+         */
+        void updateMetaData(IndexMetaData indexMetaData);
+
+        /**
+         * Checks if index requires refresh from master.
+         */
+        boolean requiresIndexMappingRefresh(IndexMetaData indexMetaData) throws IOException;
+
+        /**
+         * Returns shard with given id.
+         */
+        @Nullable T getShardOrNull(int shardId);
+
+        /**
+         * Removes shard with given id.
+         */
+        void removeShard(int shardId, String message);
+    }
+
+    public interface AllocatedIndices<T extends Shard, U extends AllocatedIndex<T>> extends Iterable<U> {
+
+        /**
+         * Returns true if changes are allowed to be made to list of indices and shards.
+         */
+        boolean changesAllowed();
```

Comment:
do we really need this? we start IndicesService before IndicesClusterStateService service and can assert this in IndicesClusterStateService.doStart method?